File: app/src/main/java/com/binderpooldemo/client/BinderPoolTool.java
package com.binderpooldemo.client;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.IBinder;
import android.os.RemoteException;
import android.util.Log;

import com.servicedemo.server.IBinderPool;

public class BinderPoolTool {

    private Context c;
    private static BinderPoolTool INSTANCE;
    private IBinderPool pool;

    private ServiceConnection mBinderConnection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            Log.e("lzw","onServiceConnected");
            pool = IBinderPool.Stub.asInterface(service);
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
            Log.e("lzw","onServiceDisconnected");
        }
    };

    private BinderPoolTool(Context c){
        this.c = c.getApplicationContext();
        bindAidl();
    }

    public static BinderPoolTool getInstance(Context c){
        if(INSTANCE == null){
            synchronized (ServiceConnection.class){
                if(INSTANCE == null){
                    INSTANCE = new BinderPoolTool(c);
                }
            }

        }
        return INSTANCE;
    }

    private void bindAidl(){
        Intent i = new Intent("com.binderpool");
        i.setPackage("com.servicedemo.server");
        c.bindService(i,mBinderConnection,Context.BIND_AUTO_CREATE);
    }

    public IBinder getBinder(int type){
        IBinder binder = null;
        try {
            binder = pool.getBinder(type);
        } catch (RemoteException e) {
            e.printStackTrace();
        }
        return binder;
    }

}

File: client/src/main/java/th/co/digio/client/BinderPoolTool.kt
package th.co.digio.client

import android.annotation.SuppressLint
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.IBinder
import android.os.RemoteException
import android.util.Log
import th.co.digio.aidl_module.IBinderPool

//import com.servicedemo.server.IBinderPool
//import com.servicedemo.server.IBinderPool.Stub

class BinderPoolTool private constructor(c: Context) {
    private val context: Context = c.applicationContext
    private var pool: IBinderPool? = null
    private val mBinderConnection: ServiceConnection = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName?, service: IBinder?) {
            Log.e("lzw", "onServiceConnected")
            pool = IBinderPool.Stub.asInterface(service)
        }

        override fun onServiceDisconnected(name: ComponentName?) {
            Log.e("lzw", "onServiceDisconnected")
        }
    }

    private fun bindAidl() {
        val i = Intent("com.binderpool")
        i.setPackage("th.co.digio.service")
        context.bindService(i, mBinderConnection, Context.BIND_AUTO_CREATE)
    }

    fun getBinder(type: Int): IBinder? {
        var binder: IBinder? = null
        try {
            binder = pool!!.getBinder(type)
        } catch (e: RemoteException) {
            e.printStackTrace()
        }
        return binder
    }

    companion object {
        @SuppressLint("StaticFieldLeak")
        private var INSTANCE: BinderPoolTool? = null

        fun getInstance(c: Context): BinderPoolTool? {
            if (INSTANCE == null) {
                synchronized(ServiceConnection::class.java) {
                    if (INSTANCE == null) {
                        INSTANCE = BinderPoolTool(c)
                    }
                }
            }
            return INSTANCE
        }
    }

    init {
        bindAidl()
    }
}